core/primitives/src/shard_layout.rs

```diff
@@ -63,7 +67,27 @@ impl ShardLayout {
         shards_split_map: Option<ShardSplitMap>,
         version: ShardVersion,
     ) -> Self {
-        Self::V1(ShardLayoutV1 { fixed_shards, boundary_accounts, shards_split_map, version })
+        let to_parent_shard_map = if let Some(shards_split_map) = &shards_split_map {
+            let mut to_parent_shard_map = HashMap::new();
+            let num_shards = (fixed_shards.len() + boundary_accounts.len() + 1) as NumShards;
+            for (parent_shard_id, shard_ids) in shards_split_map.iter().enumerate() {
+                for &shard_id in shard_ids {
+                    let prev = to_parent_shard_map.insert(shard_id, parent_shard_id as ShardId);
+                    assert!(prev.is_none(), "no shard should appear in the map twice");
+                    assert!(shard_id < num_shards, "shard id should be valid");
+                }
+            }
+            Some((0..num_shards).map(|shard_id| to_parent_shard_map[&shard_id]).collect())
```

Comment:
Maybe we should also have some informative expect here and not just panic if info for some shard is missing?